src/client/mod.rs

```diff
@@ -478,23 +480,113 @@ header! {
     (XVaultWrapTTL, "X-Vault-Wrap-TTL") => [String]
 }
 
+#[derive(Deserialize)]
+#[allow(dead_code)]         // We need only one of the key fields
+struct KeyData {
+    root_token: String,
+    keys_base64: Vec<String>,
+    keys: Vec<String>,
+}
+
+/// Represents the state of the vault server.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum VaultStatus {
+    /// The vault has been initialized, but it's in a sealed state.
+    /// It has to be unsealed with the unseal keys for use.
+    Sealed,
+    /// The vault hasn't been initialized yet.
+    Uninitialized,
+    /// The vault is unsealed. It's ready for use.
+    Unsealed,
+}
+
 impl VaultClient<TokenData> {
-    /// Construct a `VaultClient` from an existing vault token
-    pub fn new<U, T: Into<String>>(host: U, token: T) -> Result<VaultClient<TokenData>>
+    /// Construct a client by initializing a vault server.
+    #[allow(unused_results)]
+    pub fn init<U>(host: U, shares: usize, threshold: usize) -> Result<VaultClient<TokenData>>
         where U: TryInto<Url, Err = Error>
     {
-        let host = try!(host.try_into());
+        let host = host.try_into()?;
+        let client = Client::new()?;
+        let payload = json!({"secret_shares": shares, "secret_threshold": threshold});
+        let res = handle_hyper_response(client.put(host.join("/v1/sys/init")?)
+                                              .body(payload.to_string().as_str())
+                                              .send())?;
+        let data: KeyData = parse_vault_response(res)?;
+        VaultClient::new(host, data.root_token.as_str(), &data.keys)
+    }
+
+    /// Query the vault server for its status.
+    pub fn get_status<U>(host: U) -> Result<VaultStatus>
+        where U: TryInto<Url, Err = Error>
+    {
+        let host = host.try_into()?;
         let client = Client::new()?;
-        let token = token.into();
-        let res = try!(
-            handle_hyper_response(client.get(try!(host.join("/v1/auth/token/lookup-self")))
-                                  .header(XVaultToken(token.clone()))
-                                  .send()));
+        let res = handle_hyper_response(client.get(host.join("/v1/sys/init")?).send())?;
+        let decoded: Value = parse_vault_response(res)?;
+        let init = decoded.get("initialized").and_then(|v| v.as_bool())
```

Comment:
I think it would be better to implement `Deserialize` for `VaultStatus` rather than baking in the deserialize logic inside `get_status`.